include/slb/type_traits.hpp

```diff
@@ -1109,15 +1155,33 @@ using common_type_t = typename slb::common_type<Ts...>::type;
 template <typename T>
 using underlying_type_t = typename slb::underlying_type<T>::type;
 
-#if __cpp_lib_is_invocable // C++17
+// We only enable the C++17 implementation under C++2a here to account for
+// P0704: "Fixing const-qualified pointers to members".
+
+#if __cpp_lib_is_invocable /* C++17 */ && __cplusplus > 201703L /* C++2a */
 using std::invoke_result;
 using std::invoke_result_t;
 #else
-// TODO
-// template <typename F, typename... Args> struct invoke_result;
+namespace detail {
+
+template <typename F, typename Enable = void>
+struct result_of {};
+
+template <typename F, typename... Args>
+struct result_of<F(Args...),
+                 decltype((void)detail::invoke(std::declval<F>(),
+                                               std::declval<Args>()...))> {
+  using type =
+      decltype(detail::invoke(std::declval<F>(), std::declval<Args>()...));
+};
+
+} // namespace detail
+
+template <typename Fn, typename... ArgTypes>
+struct invoke_result : detail::result_of<Fn && (ArgTypes && ...)> {};
```

Comment:
Is the function type formulation safe as long as the `F` and `Args` have references on them?
On a glance it seems to defeat the purpose of `invoke_result`.

Reply:
Yes, the function formulation is safe.

The purpose of `invoke_result` is to present the user with a less error prone interface.